scripts/scil_compute_local_tracking.py

```diff
@@ -230,61 +374,49 @@ def main():
-        random_seed=args.seed,
-        save_seeds=args.save_seeds)
 
-    scaled_min_length = args.min_length / voxel_size
-    scaled_max_length = args.max_length / voxel_size
-
-    if args.save_seeds:
-        filtered_streamlines, seeds = \
-            zip(*((s, p) for s, p in tqdm_if_verbose(
-                streamlines_generator, verbose=args.verbose,
-                total=total_nb_seeds, miniters=int(total_nb_seeds / 100))
-                  if scaled_min_length <= length(s) <= scaled_max_length))
-        data_per_streamlines = {'seeds': lambda: seeds}
-    else:
-        filtered_streamlines = \
-            (s for s in tqdm_if_verbose(
-                streamlines_generator, verbose=args.verbose,
-                total=total_nb_seeds, miniters=int(total_nb_seeds / 100))
-             if scaled_min_length <= length(s) <= scaled_max_length)
-        data_per_streamlines = {}
-
-    if args.compress:
-        # Compressing. Threshold is in mm, but we are working in voxel space.
-        # Equivalent of sft.to_voxmm:  streamline *= voxres
-        # Equivalent of sft.to_vox: streamline /= voxres
-        voxres = np.asarray(odf_sh_img.header.get_zooms()[0:3])
-        filtered_streamlines = (
-            compress_streamlines(s * voxres,
-                                 args.compress) / voxres
-            for s in filtered_streamlines)
-
-    tractogram = LazyTractogram(lambda: filtered_streamlines,
-                                data_per_streamlines,
-                                affine_to_rasmm=seed_img.affine)
-
-    filetype = nib.streamlines.detect_format(args.out_tractogram)
-    reference = get_reference_info(seed_img)
-    header = create_tractogram_header(filetype, *reference)
-
-    # Use generator to save the streamlines on-the-fly
-    nib.streamlines.save(tractogram, args.out_tractogram, header=header)
+    # NOTE: tracking is performed in voxel space in both cases
+    if not args.use_gpu:
+        # LocalTracking.maxlen is actually the maximum length per direction
+        # so we need to divide by 2.0 to limit the streamline to the right
+        # length
+        max_steps = int(args.max_length / 2.0 / args.step_size)
+        total_nb_seeds = len(seeds)
+
+        streamlines_generator = LocalTracking(
+            _get_direction_getter(args),
+            BinaryStoppingCriterion(mask_data),
+            seeds, np.eye(4),
+            step_size=vox_step_size, max_cross=1,
+            maxlen=max_steps,
+            fixedstep=True, return_all=True,
+            random_seed=args.seed,
+            save_seeds=args.save_seeds)
+        save_tractogram_cpu(streamlines_generator, odf_sh_img,
+                            seed_img, total_nb_seeds, args)
+    else:  # GPU tracking
+        vox_max_length = args.max_length / voxel_size
+        vox_min_length = args.min_length / voxel_size
+        min_strl_len = int(vox_min_length / vox_step_size) + 1
+        max_strl_len = int(vox_max_length / vox_step_size) + 1
```

Comment:
For a 0.5 mm difference: I think your file local_tracking.cl should have `while(current_length <= max_length && is_valid)` rather than `<` at line 283.

If I  understood right:

For a voxel size of 1x1x1, length 20-200, step size 0.5:

IN CPU:
           - `save_tractogram_cpu` keeps lines if  `min <= length_mm <= max`
           - Where min = 20mm, max=200mm (`scaled_min_length = args.min_length / voxel_size`) are included.

IN GPU:
           - In tracker.py, line 577, we have (MAX_LENGTH=max_strl_points)
           - Where min = 41 points, max = 401 pts.
           - And then while(current_length < max_length). So if I understand you stop at current_length=400 pts.
           - So final length is 400 points = 399 segments of 0.5mm = 199.5mm.

It took me to read very far to find all that information. Maybe add a few comments here in the main?

Reply:
For GPU tracking the point is added before incrementing, and therefore 401th point is added before we break out of the `while` loop. Also I tested on the fibercup, setting a small maximum length for both CPU and GPU. mi-brain reports the right Maximum length in both cases.